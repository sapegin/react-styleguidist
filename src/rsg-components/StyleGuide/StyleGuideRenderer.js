import React, { PropTypes } from 'react';
import Logo from 'rsg-components/Logo';
import Markdown from 'rsg-components/Markdown';
import Styled from 'rsg-components/Styled';
import cx from 'classnames';

const styles = ({ font, base, light, border, baseBackground, codeBackground }) => ({
	root: {
		isolate: false,
		color: base,
		backgroundColor: baseBackground,
	},
	hasSidebar: {
		isolate: false,
		paddingLeft: 200,
	},
	content: {
		isolate: false,
		maxWidth: 1000,
		padding: [[15, 30]],
		margin: [[0, 'auto']],
	},
	sidebar: {
		backgroundColor: codeBackground,
		borderRight: [[1, border, 'solid']],
		position: 'fixed',
		top: 0,
		left: 0,
		bottom: 0,
		width: 200,
		overflow: 'scroll',
	},
	components: {
		isolate: false,
		overflow: 'auto',  // To prevent the pane from growing out of the screen
	},
	logo: {
		padding: 15,
		borderBottom: [[1, border, 'solid']],
	},
	footer: {
		color: light,
		fontFamily: font,
		fontSize: 12,
	},
});

export const StyleGuideRenderer = ({ classes, title, homepageUrl, components, toc, sidebar }) => (
	<div className={cx(classes.root, sidebar && classes.hasSidebar)}>
		<main className={classes.content}>
			<div className={classes.components}>
				{components}
				<footer className={classes.footer}>
					<Markdown text={`Generated with [React Styleguidist](${homepageUrl})`} />
				</footer>
			</div>
		</main>
		{sidebar &&
			<div className={classes.sidebar}>
				<div className={classes.logo}>
					<Logo>{title}</Logo>
				</div>
				{toc}
			</div>
		}
	</div>
);

StyleGuideRenderer.propTypes = {
	classes: PropTypes.object.isRequired,
	title: PropTypes.string.isRequired,
	homepageUrl: PropTypes.string.isRequired,
	components: PropTypes.object.isRequired,
	toc: PropTypes.node.isRequired,
	sidebar: PropTypes.bool,
};

export default Styled(styles)(StyleGuideRenderer);
