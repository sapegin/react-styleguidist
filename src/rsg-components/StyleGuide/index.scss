.main {
  position: relative;
  max-width: 200000px;
  padding: 0px 0px 46px 32px;
  margin: 0 auto;
  display: block;
}

.footer {
  width: 100%;
  position: absolute;
  bottom: 0;
  left: 0;
  background: whitesmoke;
  padding: 12px 32px;
  border: 1px solid #e8e8e8;
  border-width: 1px 0px 0px 0px;

  p {
	margin: 0;
  }
}
