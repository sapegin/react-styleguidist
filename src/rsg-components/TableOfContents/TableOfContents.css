.root {
	composes: reset font from "../../styles/common.css";
}

.list {
	margin: 0;
	padding: 15px;
	list-style-type: none;
}

.item {
	margin: 0 0 7px;
	padding: 0;
	font-size: 15px;
}

.link {
	composes: link from "../../styles/colors.css";
}

.section {
	composes: link from "../../styles/colors.css";
	font-weight: bold;
}

.search {
	composes: base base-bg border from "../../styles/colors.css";
	display: block;
	box-sizing: border-box;
	width: 170px;
	margin: 15px 0 0 15px;
	padding: 6px 12px;
	font-size: 15px;
	border-radius: 2px;
	transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
}
.search:focus {
	border-color: #1978c8;
	outline: 0;
}

.isolatedLink {
	composes: link from "../../styles/colors.css";
	margin-left: 5px;
	display: none;
	position: absolute;
}

.item:hover .isolatedLink {
	display: inline;
}
