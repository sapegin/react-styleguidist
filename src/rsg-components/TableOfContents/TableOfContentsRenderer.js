import React, { PropTypes } from 'react';
import Styled from 'rsg-components/Styled';

const styles = ({ font, base, border, link, baseBackground }) => ({
	root: {
		fontFamily: font,
	},
	search: {
		padding: 15,
	},
	input: {
		display: 'block',
		width: '100%',
		padding: [[6, 12]],
		color: base,
		backgroundColor: baseBackground,
		fontFamily: font,
		fontSize: 15,
		border: [[1, border, 'solid']],
		borderRadius: 2,
		transition: 'border-color ease-in-out .15s',
		'&:focus': {
			isolate: false,
			borderColor: link,
			outline: 0,
		},
	},
});

export function TableOfContentsRenderer({
	classes,
	children,
	searchTerm,
	onSearchTermChange,
}) {
	return (
		<div>
			<div className={classes.root}>
				<div className={classes.search}>
					<input
						value={searchTerm}
						className={classes.input}
						placeholder="Filter by name"
						onChange={event => onSearchTermChange(event.target.value)}
					/>
				</div>
				{children}
			</div>
		</div>
	);
}

TableOfContentsRenderer.propTypes = {
	classes: PropTypes.object.isRequired,
	children: PropTypes.node,
	searchTerm: PropTypes.string.isRequired,
	onSearchTermChange: PropTypes.func.isRequired,
};

export default Styled(styles)(TableOfContentsRenderer);
