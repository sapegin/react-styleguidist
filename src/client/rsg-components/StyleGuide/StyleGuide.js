import React, { Component } from 'react';
import PropTypes from 'prop-types';
import TableOfContents from 'rsg-components/TableOfContents';
import StyleGuideRenderer from 'rsg-components/StyleGuide/StyleGuideRenderer';
import Sections from 'rsg-components/Sections';
import Welcome from 'rsg-components/Welcome';
import Error from 'rsg-components/Error';
import NotFound from 'rsg-components/NotFound';
import Context from 'rsg-components/Context';
import { HOMEPAGE } from '../../../scripts/consts';
import { DisplayModes } from '../../consts';

/**
 * This function will return true, if the sidebar should be visible and false otherwise.
 *
 * These sorted conditions (highest precedence first) define the visibility
 * state of the sidebar.
 *
 * - Sidebar is hidden for isolated example views
 * - Sidebar is always visible when pagePerSection
 * - Sidebar is hidden when showSidebar is set to false
 * - Sidebar is visible when showSidebar is set to true for non-isolated views
 *
 * @param {boolean} displayMode
 * @param {boolean} showSidebar
 * @param {boolean} pagePerSection
 * @returns {boolean}
 */
function hasSidebar(displayMode, showSidebar) {
	return displayMode === DisplayModes.notFound || (showSidebar && displayMode === DisplayModes.all);
}

export default class StyleGuide extends Component {
	static propTypes = {
		codeRevision: PropTypes.number.isRequired,
		config: PropTypes.object.isRequired,
		slots: PropTypes.object.isRequired,
		sections: PropTypes.array.isRequired,
		welcomeScreen: PropTypes.bool,
		patterns: PropTypes.array,
		displayMode: PropTypes.string,
		allSections: PropTypes.array.isRequired,
		pagePerSection: PropTypes.bool,
	};
	static defaultProps = {
		displayMode: DisplayModes.all,
	};

	state = {
		error: false,
		info: null,
	};

	componentDidCatch(error, info) {
		this.setState({
			error,
			info,
		});
	}

	render() {
		const { error, info } = this.state;
		const {
			config,
			sections,
			welcomeScreen,
			patterns,
			displayMode,
			allSections,
			pagePerSection,
			codeRevision,
			slots,
		} = this.props;

		if (error) {
			return <Error error={error} info={info} />;
		}

		if (welcomeScreen) {
			return <Welcome patterns={patterns} />;
		}

		return (
			<Context.Provider
				value={{
					codeRevision,
					config,
					slots,
					displayMode,
				}}
			>
				<StyleGuideRenderer
					title={config.title}
					version={config.version}
					homepageUrl={HOMEPAGE}
					toc={<TableOfContents sections={allSections} useRouterLinks={pagePerSection} />}
					hasSidebar={hasSidebar(displayMode, config.showSidebar)}
				>
					{sections.length ? <Sections sections={sections} depth={1} /> : <NotFound />}
				</StyleGuideRenderer>
			</Context.Provider>
		);
	}
}
