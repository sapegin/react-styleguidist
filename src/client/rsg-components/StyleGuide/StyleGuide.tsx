import React, { Component } from 'react';
import PropTypes from 'prop-types';
import TableOfContents from 'rsg-components/TableOfContents';
import StyleGuideRenderer from 'rsg-components/StyleGuide/StyleGuideRenderer';
import Sections from 'rsg-components/Sections';
import Welcome from 'rsg-components/Welcome';
import Error from 'rsg-components/Error';
import NotFound from 'rsg-components/NotFound';
import Context from 'rsg-components/Context';
import MdxProvider from 'rsg-components/mdx/MdxProvider';
import { HOMEPAGE } from '../../../scripts/consts';
import * as Rsg from '../../../typings';

/**
 * This function will return true, if the sidebar should be visible and false otherwise.
 *
 * These sorted conditions (highest precedence first) define the visibility
 * state of the sidebar.
 *
 * - Sidebar is always hidden in isolated mode
 * - Sidebar is visible when pagePerSection
 * - Sidebar is hidden when showSidebar is set to false
 * - Sidebar is visible when showSidebar is set to true for non-isolated views
 */
function hasSidebar(sections: Rsg.Section[], isolated: boolean, showSidebar: boolean): boolean {
	return sections.length > 0 && !isolated && showSidebar;
}

export interface StyleGuideProps {
	codeRevision: number;
	cssRevision: string;
	config: Rsg.ProcessedStyleguidistConfig;
	slots: any;
	sections: Rsg.Section[];
	welcomeScreen?: boolean;
	patterns?: string[];
	isolated: boolean;
	exampleIndex?: number | string;
	allSections?: Rsg.Section[];
	pagePerSection?: boolean;
}

interface StyleGuideState {
	error: Error | boolean;
	info: React.ErrorInfo | null;
}

export default class StyleGuide extends Component<StyleGuideProps, StyleGuideState> {
	public static propTypes = {
		codeRevision: PropTypes.number.isRequired,
		cssRevision: PropTypes.string.isRequired,
		config: PropTypes.object.isRequired,
		slots: PropTypes.object.isRequired,
		sections: PropTypes.array.isRequired,
		welcomeScreen: PropTypes.bool,
		patterns: PropTypes.array,
		isolated: PropTypes.bool.isRequired,
		exampleIndex: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
		allSections: PropTypes.array.isRequired,
		pagePerSection: PropTypes.bool,
	};

	public state = {
		error: false,
		info: null,
	};

	public componentDidCatch(error: Error, info: React.ErrorInfo) {
		this.setState({
			error,
			info,
		});
	}

	public render() {
		const { error, info }: StyleGuideState = this.state;
		const {
			config,
			sections,
			welcomeScreen,
			patterns,
			isolated,
			exampleIndex,
			allSections,
			pagePerSection,
			codeRevision,
			cssRevision,
			slots,
		} = this.props;

		if (error && info) {
			return <Error error={error} info={info} />;
		}

		if (welcomeScreen && patterns) {
			return <Welcome patterns={patterns} />;
		}

		return (
			<Context.Provider
				value={{
					codeRevision,
					config,
					slots,
					isolated,
					exampleIndex,
					cssRevision,
				}}
			>
				<MdxProvider>
					<StyleGuideRenderer
						key={cssRevision}
						title={config.title}
						version={config.version}
						homepageUrl={HOMEPAGE}
						toc={
							allSections ? (
								<TableOfContents
									sections={allSections}
									pagePerSection={pagePerSection}
									tocMode={config.tocMode}
								/>
							) : null
						}
						hasSidebar={hasSidebar(sections, isolated, config.showSidebar)}
					>
						{sections.length ? <Sections sections={sections} depth={1} /> : <NotFound />}
					</StyleGuideRenderer>
				</MdxProvider>
			</Context.Provider>
		);
	}
}
