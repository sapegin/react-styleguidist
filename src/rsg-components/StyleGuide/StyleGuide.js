import React, { Component } from 'react';
import PropTypes from 'prop-types';
import TableOfContents from 'rsg-components/TableOfContents';
import StyleGuideRenderer from 'rsg-components/StyleGuide/StyleGuideRenderer';
import Sections from 'rsg-components/Sections';
import Welcome from 'rsg-components/Welcome';
import { HOMEPAGE } from '../../../scripts/consts';
import { filterSectionsByName } from '../../utils/utils';

export default class StyleGuide extends Component {
	static propTypes = {
		codeKey: PropTypes.number.isRequired,
		config: PropTypes.object.isRequired,
		slots: PropTypes.object.isRequired,
		sections: PropTypes.array.isRequired,
		welcomeScreen: PropTypes.bool,
		patterns: PropTypes.array,
		isolatedComponent: PropTypes.bool,
		isolatedExample: PropTypes.bool,
		isolatedSection: PropTypes.bool,
	};

	static childContextTypes = {
		codeKey: PropTypes.number.isRequired,
		config: PropTypes.object.isRequired,
		slots: PropTypes.object.isRequired,
		isolatedComponent: PropTypes.bool,
		isolatedExample: PropTypes.bool,
		isolatedSection: PropTypes.bool,
	};

	static defaultProps = {
		isolatedComponent: false,
	};

	state = {
		searchTerm: '',
	};

	getChildContext() {
		return {
			codeKey: this.props.codeKey,
			config: this.props.config,
			slots: this.props.slots,
			isolatedComponent: this.props.isolatedComponent,
			isolatedExample: this.props.isolatedExample,
			isolatedSection: this.props.isolatedSection,
		};
	}

	render() {
		const { config, sections, welcomeScreen, patterns, isolatedComponent } = this.props;

		const { searchTerm } = this.state;
		const filteredSections = (0, filterSectionsByName)(sections, searchTerm);

		if (welcomeScreen) {
			return <Welcome patterns={patterns} />;
		}

		return (
			<StyleGuideRenderer
				title={config.title}
				homepageUrl={HOMEPAGE}
				hasSidebar={config.showSidebar && !isolatedComponent}
				toc={
					<TableOfContents
						sections={filteredSections}
						searchTerm={searchTerm}
						onSearchTermChange={searchTerm => this.setState({ searchTerm })}
					/>
				}
			>
				<Sections sections={filteredSections} />
			</StyleGuideRenderer>
		);
	}
}
