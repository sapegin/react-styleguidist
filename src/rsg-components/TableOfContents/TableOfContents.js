import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ComponentsList from 'rsg-components/ComponentsList';
import TableOfContentsRenderer from 'rsg-components/TableOfContents/TableOfContentsRenderer';
import filterSectionsByName from '../../utils/filterSectionsByName';

export default class TableOfContents extends Component {
	static propTypes = {
		sections: PropTypes.array.isRequired,
	};
	state = {
		searchTerm: '',
	};

	renderLevel(sections, oneComponentPerPage) {
		const items = sections.map(section => {
			const children = [...(section.sections || []), ...(section.components || [])];
			return Object.assign({}, section, {
				heading: !!section.name && children.length > 0,
				content: children.length > 0 && this.renderLevel(children),
			});
		});
		return <ComponentsList items={items} oneComponentPerPage={oneComponentPerPage} />;
	}

	renderSections() {
		const { searchTerm } = this.state;
		const { sections } = this.props;
		const oneComponentPerPage = (this.context.config || {}).oneComponentPerPage;
		// If there is only one section, we treat it as a root section
		// In this case the name of the section won't be rendered and it won't get left padding
		const firstLevel = sections.length === 1 ? sections[0].components : sections;
		const filtered = filterSectionsByName(firstLevel, searchTerm);

		return this.renderLevel(filtered, oneComponentPerPage);
	}

	render() {
		const { searchTerm } = this.state;
		return (
			<TableOfContentsRenderer
				searchTerm={searchTerm}
				onSearchTermChange={searchTerm => this.setState({ searchTerm })}
			>
				{this.renderSections()}
			</TableOfContentsRenderer>
		);
	}
}
