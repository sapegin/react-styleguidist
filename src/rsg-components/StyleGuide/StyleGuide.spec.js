import React from 'react';
import TableOfContents from '../TableOfContents';
import StyleGuide from './StyleGuide';
import { StyleGuideRenderer } from './StyleGuideRenderer';

const sections = [
	{
		components: [
			{
				name: 'Foo',
				pathLine: 'components/foo.js',
				filepath: 'components/foo.js',
				props: {
					description: 'Foo foo',
				},
			},
			{
				name: 'Bar',
				pathLine: 'components/bar.js',
				filepath: 'components/bar.js',
				props: {
					description: 'Bar bar',
				},
			},
		],
	},
];
const config = {
	title: 'Hello',
	showSidebar: true,
};

it('should render components list', () => {
	const actual = shallow(<StyleGuide codeKey={1} config={config} sections={sections} />);

	expect(actual).toMatchSnapshot();
});

it('should render welcome screen', () => {
	const actual = shallow(<StyleGuide codeKey={1} config={config} sections={[]} welcomeScreen />);

	expect(actual).toMatchSnapshot();
});

describe('sidebar rendering', () => {
	it('renderer should have sidebar if showSidebar is not set', () => {
		const wrapper = shallow(<StyleGuide codeKey={1} config={config} sections={sections} />);

		expect(wrapper.prop('hasSidebar')).toEqual(true);
	});

	it('renderer should not have sidebar if showSidebar is false', () => {
		const wrapper = shallow(
			<StyleGuide
				codeKey={1}
				config={{
					...config,
					showSidebar: false,
				}}
				sections={sections}
			/>
		);

		expect(wrapper.prop('hasSidebar')).toEqual(false);
	});

	it('renderer should not have sidebar in isolation mode', () => {
		const wrapper = shallow(
			<StyleGuide codeKey={1} config={config} sections={sections} isolatedComponent />
		);

		expect(wrapper.prop('hasSidebar')).toEqual(false);
	});
});

it('renderer should render logo, table of contents and passed children', () => {
	const actual = shallow(
		<StyleGuideRenderer
			classes={{}}
			title={config.title}
			toc={<TableOfContents sections={sections} />}
			homepageUrl="http://react-styleguidist.js.org/"
			hasSidebar
		>
			<h1>Content</h1>
		</StyleGuideRenderer>
	);

	expect(actual).toMatchSnapshot();
});
