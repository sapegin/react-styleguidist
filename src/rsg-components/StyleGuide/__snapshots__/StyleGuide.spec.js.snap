// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer should render logo, version, table of contents, ribbon and passed children 1`] = `
<div
  className=""
>
  <main>
    <h1>
      Content
    </h1>
    <footer>
      <Markdown
        text="Generated with [React Styleguidist](http://react-styleguidist.js.org/)"
      />
    </footer>
  </main>
  <div>
    <div>
      <Consumer(Logo)>
        Hello
      </Consumer(Logo)>
      <Consumer(Version)>
        1.0.0
      </Consumer(Version)>
    </div>
    <TableOfContents
      sections={
        Array [
          Object {
            "components": Array [
              Object {
                "filepath": "components/foo.js",
                "name": "Foo",
                "pathLine": "components/foo.js",
                "props": Object {
                  "description": "Foo foo",
                },
              },
              Object {
                "filepath": "components/bar.js",
                "name": "Bar",
                "pathLine": "components/bar.js",
                "props": Object {
                  "description": "Bar bar",
                },
              },
            ],
          },
        ]
      }
    />
  </div>
  <Consumer(Ribbon) />
</div>
`;

exports[`should render an error when componentDidCatch() is triggered 1`] = `
<Consumer(Error)
  error={
    Object {
      "toString": [Function],
    }
  }
  info={
    Object {
      "componentStack": Object {
        "toString": [Function],
      },
    }
  }
/>
`;

exports[`should render components list 1`] = `
<Provider
  codeRevision={1}
  config={
    Object {
      "showSidebar": true,
      "title": "Hello",
      "version": "1.0.0",
    }
  }
  displayMode="all"
  slots={Object {}}
>
  <Consumer(StyleGuide)
    hasSidebar={true}
    homepageUrl="https://react-styleguidist.js.org/"
    title="Hello"
    toc={
      <TableOfContents
        sections={
          Array [
            Object {
              "components": Array [
                Object {
                  "filepath": "components/foo.js",
                  "name": "Foo",
                  "pathLine": "components/foo.js",
                  "props": Object {
                    "description": "Foo foo",
                  },
                },
                Object {
                  "filepath": "components/bar.js",
                  "name": "Bar",
                  "pathLine": "components/bar.js",
                  "props": Object {
                    "description": "Bar bar",
                  },
                },
              ],
            },
          ]
        }
        useRouterLinks={false}
      />
    }
    version="1.0.0"
  >
    <Sections
      depth={1}
      sections={
        Array [
          Object {
            "components": Array [
              Object {
                "filepath": "components/foo.js",
                "name": "Foo",
                "pathLine": "components/foo.js",
                "props": Object {
                  "description": "Foo foo",
                },
              },
              Object {
                "filepath": "components/bar.js",
                "name": "Bar",
                "pathLine": "components/bar.js",
                "props": Object {
                  "description": "Bar bar",
                },
              },
            ],
          },
        ]
      }
    />
  </Consumer(StyleGuide)>
</Provider>
`;

exports[`should render welcome screen 1`] = `<Consumer(Welcome) />`;
